Server startup may fail unexpectedly

### Issue description

It looks like in some cases, starting may fail with this exception:

```


Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0 |  
-- | --


	at io.micronaut.testresources.buildtools.ServerUtils.startOrConnectToExistingServer(ServerUtils.java:179)

```

Which is probably caused by the fact that the server file is present, but not flushed yet.

--- a/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
+++ b/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
@@ -59,6 +59,7 @@ public class ServerUtils {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(ServerUtils.class.getName());
     private static final int STARTUP_TIME_WAIT_MS = 200;
+    private static final int MAX_READS = 10;
 
     private static final String SERVER_URI = "server.uri";
     private static final String SERVER_ACCESS_TOKEN = "server.access.token";
@@ -176,7 +177,22 @@ public class ServerUtils {
         startAndWait(serverFactory, explicitPort, portFilePath, accessToken, serverClasspath, cdsDirectory);
         int port;
         if (explicitPort == null) {
-            port = Integer.parseInt(Files.readAllLines(portFilePath).get(0));
+            List<String> lines = Files.readAllLines(portFilePath);
+            int attempts = 1;
+            while (lines.isEmpty()) {
+                if (attempts == MAX_READS) {
+                    throw new IllegalStateException("Unable to read port file " + portFilePath + ": file is empty");
+                }
+                // It is still possible to see the file, but that its contents isn't flushed yet
+                try {
+                    serverFactory.waitFor(Duration.of(STARTUP_TIME_WAIT_MS, ChronoUnit.MILLIS));
+                    lines = Files.readAllLines(portFilePath);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                attempts++;
+            }
+            port = Integer.parseInt(lines.get(0));
         } else {
             port = explicitPort;
         }

--- a/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
+++ b/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
@@ -212,6 +212,60 @@ class ServerUtilsTest extends Specification {
         applicationContext.stop()
     }
 
+    def "waits for port file to have contents"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def applicationContext = ApplicationContext.builder().start()
+        def embeddedServer = applicationContext.getBean(EmbeddedServer)
+        embeddedServer.start()
+
+        when:
+        def settings = ServerUtils.startOrConnectToExistingServer(null, portFile, settingsDir, null, [], null, factory)
+
+        then:
+        1 * factory.startServer(_) >> { ServerUtils.ProcessParameters params ->
+
+        }
+        1 * factory.waitFor(_) >> {
+            portFile.toFile().text = ""
+            println "Waiting to write the contents of the port file"
+        }
+        1 * factory.waitFor(_) >> {
+            portFile.toFile().text = "${embeddedServer.port}"
+        }
+
+        cleanup:
+        applicationContext.stop()
+    }
+
+    def "reasonable error message if port file can never be read"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def applicationContext = ApplicationContext.builder().start()
+        def embeddedServer = applicationContext.getBean(EmbeddedServer)
+        embeddedServer.start()
+
+        when:
+        def settings = ServerUtils.startOrConnectToExistingServer(null, portFile, settingsDir, null, [], null, factory)
+
+        then:
+        1 * factory.startServer(_) >> { ServerUtils.ProcessParameters params ->
+
+        }
+        10 * factory.waitFor(_) >> {
+            portFile.toFile().text = ""
+            println "Waiting to write the contents of the port file"
+        }
+        IllegalStateException ex = thrown()
+        ex.message == "Unable to read port file ${portFile}: file is empty"
+
+        cleanup:
+        applicationContext.stop()
+    }
+
+
     @Controller
     static class ServerMock {
 

